jdaviz/configs/default/plugins/model_fitting/model_fitting.py

```diff
@@ -800,23 +808,41 @@ def _register_spectrum(self, event):
         self.add_results.add_results_from_plugin(spectrum)
         self._set_default_results_label()
 
-    def _apply_subset_mask(self, spectrum):
-        """"
-        Apply mask from selected subset to ``spectrum``
+    def _apply_subset_masks(self, spectrum, dimensions=['spectral', 'spatial']):
         """
-        if self.spectral_subset_selected != "Entire Spectrum":
-            subset_mask = self.app.get_data_from_viewer(
-                self._default_spectrum_viewer_reference_name,
-                data_label=self.spectral_subset_selected
-            ).mask
+        Iterate over dimensions of a spectral cube and mask out the
+        subsets along each axis
+        """
+        for dimension in dimensions:
+            if dimension == 'spectral' and hasattr(self, '_default_spectrum_viewer_reference_name'):
+                viewer_ref = self._default_spectrum_viewer_reference_name
+                data_label = self.spectral_subset_selected
+            elif dimension == 'spatial' and hasattr(self, '_default_flux_viewer_reference_name'):
+                viewer_ref = self._default_flux_viewer_reference_name
+                data_label = self.spatial_subset_selected
+            else:
+                # If this viewer isn't available, don't try to apply the subset mask
+                continue
 
-            if subset_mask.shape != spectrum.shape:
-                # broadcast to the dimensions of the full cube:
-                subset_mask = np.broadcast_to(
-                    subset_mask[None, None, :], spectrum.shape
+            if (not getattr(self, dimension + '_subset_selected').startswith("Entire") and
+                    viewer_ref in self.app.get_viewer_reference_names()):
+                subset = self.app.get_data_from_viewer(
+                    viewer_ref, data_label=data_label
```

Comment:
can we move this up higher to `SubsetSelect.selected_subset_data` (or something like that)?  Then you can rely on the component to do all the viewer accessing as well and hide all the ugly code somewhere that is a bit more reusable.

Reply:
I think I explained this poorly.  I'm not sure moving the entire `_apply_subset_masks` will work (cleanly) because of the way model fitting is caching the spectrum object manually outside of the components.  I was thinking more of just this single line (and the lines before it to retrieve the applicable viewer).